### PR TITLE
Fixed W3C error messaging

### PR DESCRIPTION
**Pull request by dpgraham**
* Rename '.error' property to correct W3C property
    * The error property in the ProtocolError objects should match up with https://www.w3.org/TR/webdriver/#handling-errors under the 'JSON Error Code' column
    * Did this and added tests for all of these
* Add proxying for W3C responses 
    * If a W3C Response is provided by a proxy, map to the correct error
    * Checks if `value` is an object and `value.error` and maps it based on that (`value.error` is the W3C error code)
    * Broke off `errorFromCode` into `errorFromMJSONWPStatusCode` and `errorFromW3CJsonCode`

### CODE REVIEW

File: lib/protocol/errors.js [hunk 1]
```diff
@@ -533,23 +539,38 @@ class BadParametersError extends ES6Error {
...
+        origMessage = responseError.value.message;
+      }
+    }
+    let message = `Proxy request unsuccessful. ${origMessage}`;
     super(err || message);
+
     this.w3cStatus = HTTPStatusCodes.BAD_REQUEST;
-    if (typeof jsonwp === 'string') {
-      this.jsonwp = JSON.parse(jsonwp);
+    if (typeof responseError === 'string') {
```
Comment (mykola-mokhnach) [inline lib/protocol/errors.js]: _.isString

File: lib/protocol/errors.js [hunk 2]
```diff
@@ -613,17 +643,36 @@ function isErrorType (err, type) {
     let hasJsonwpObj = !!err.jsonwp;
```
Comment (mykola-mokhnach) [inline lib/protocol/errors.js]: this boolean conversion looks redundant to me

File: lib/protocol/errors.js [hunk 3]
```diff
@@ -613,17 +643,36 @@ function isErrorType (err, type) {
...
 }
 
-// retrieve an error with the code and message
-function errorFromCode (code, message) {
-  if (code !== 13 && errorCodeMap[code]) {
-    return new errorCodeMap[code](message);
+/**
+ * Retrieve an error derived from MJSONWP status
+ * @param {number} code JSONWP status code
+ * @param {string} message
```
Comment (mykola-mokhnach) [inline lib/protocol/errors.js]: `@return` description is missing

File: lib/protocol/errors.js [hunk 4]
```diff
@@ -613,17 +643,36 @@ function isErrorType (err, type) {
...
+ */
+function errorFromMJSONWPStatusCode (code, message) {
+  if (code !== 13 && jsonwpErrorCodeMap[code]) {
```
Comment (mykola-mokhnach) [inline lib/protocol/errors.js]: do we have a constant for magic number 13?

File: lib/protocol/errors.js [hunk 5]
```diff
@@ -613,17 +643,36 @@ function isErrorType (err, type) {
...
+    return new jsonwpErrorCodeMap[code](message);
+  }
+  return new UnknownError(message);
+}
+
+/**
+ * Retrieve an error derived from W3C JSON Code
+ * @param {string} code W3C error string (see https://www.w3.org/TR/webdriver/#handling-errors `JSON Error Code` column)
```
Comment (mykola-mokhnach) [inline lib/protocol/errors.js]: missing message parameter description and returned value description

File: lib/protocol/protocol.js [hunk 6]
```diff
@@ -295,8 +295,12 @@ function buildHandler (app, method, path, spec, driver, isSessCmd) {
       }
 
       // if the status is not 0,  throw the appropriate error for status code.
-      if (util.hasValue(driverRes) && util.hasValue(driverRes.status) && parseInt(driverRes.status, 10) !== 0) {
-        throw errorFromCode(driverRes.status, driverRes.value);
+      if (util.hasValue(driverRes)) {
+        if (util.hasValue(driverRes.status) && parseInt(driverRes.status, 10) !== 0) {
```
Comment (mykola-mokhnach) [inline lib/protocol/protocol.js]: I'd also check if `!isNaN(driverRes.status)` before casting it to integer

**Comment by dpgraham** [comment]
Fixes pushed

File: lib/protocol/protocol.js [hunk 7]
```diff
@@ -296,7 +296,7 @@ function buildHandler (app, method, path, spec, driver, isSessCmd) {
 
       // if the status is not 0,  throw the appropriate error for status code.
       if (util.hasValue(driverRes)) {
-        if (util.hasValue(driverRes.status) && parseInt(driverRes.status, 10) !== 0) {
+        if (util.hasValue(driverRes.status) && parseInt(driverRes.status, 10) !== 0 && !_.isNaN(driverRes.status)) {
```
Comment (mykola-mokhnach) [inline lib/protocol/protocol.js]: Use built-in javascript isNaN function rather than lodash one. Also, this check should be placed before `parseInt`

**Pull request merged**